torchmultimodal/utils/generate.py

```diff
@@ -0,0 +1,212 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```

Comment:
Why put this in utils? It feels like more of an example script

Reply:
The idea is that any GPT-like model with AR decoding feature can be plugged into this utility to generate data.
It is more of a layer of abstraction rather than a full blown script with project-specific details such as datamodule and etc. 

For example:
- Huggingface designs the similar component as a model mixin: https://github.com/huggingface/transformers/blob/v4.21.2/src/transformers/generation_utils.py#L379
- Fairseq does so by enforcing a model to inherit from `FairseqDecoder`: https://fairseq.readthedocs.io/en/latest/_modules/fairseq/models/fairseq_incremental_decoder.html

I am just taking the middle route (i.e., as a utility function) that is neither a mixin nor through inheritance.

Reply:
Thanks for the explanation. It would be nice to see a toy example of how this utility fits together with model class and other components like datamodule. But not really a concern for this PR, mainly just something that would help my own understanding

Reply:
Yeah once I load the checkpoint will show that along with the dataset.
Right now the conceptual idea can be seen from the unit test where `model` is `video_gpt` and data is random inputs:

https://github.com/facebookresearch/multimodal/blob/a896bf7b952d08957ea279c05af66c2643dbdc9b/test/utils/test_generate.py#L65-L83

Note that the patching of multinomial is needed here because the model weights are random (i.e., unconditioned on any training data) so sampling can spit out tokens that are out side of the "vocabulary". Therefore, we force sampling to return always `0` token while still being able to test the generation loop.